File: Java/bCAPClient/Src/orin2/bcap/BCAPPacketConverter.java
/**
 * Software License Agreement (MIT License)
 *
 * @copyright Copyright (c) 2015 DENSO WAVE INCORPORATED
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package orin2.bcap;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.lang.reflect.Array;
import java.math.BigDecimal;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Date;
import java.util.Hashtable;
import java.util.Map.Entry;

class BCAPPacketConverter {
	private static final int VT_EMPTY   = 0;
	private static final int VT_NULL    = 1;
	private static final int VT_ERROR   = 10;
	private static final int VT_UI1     = 17;
	private static final int VT_I2      = 2;
	private static final int VT_UI2     = 18;
	private static final int VT_I4      = 3;
	private static final int VT_UI4     = 19;
	private static final int VT_I8      = 20;
	private static final int VT_UI8     = 21;
	private static final int VT_R4      = 4;
	private static final int VT_R8      = 5;
	private static final int VT_CY	    = 6;
	private static final int VT_DATE    = 7;
	private static final int VT_BOOL    = 11;
	private static final int VT_BSTR    = 8;
	private static final int VT_VARIANT = 12;
	private static final int VT_ARRAY   = 0x2000;
	
	private static final int BCAP_SOH = 0x1;
	private static final int BCAP_EOT = 0x4;
	private static final int MAX_OPT_BYTES = 8;
	
	private static final int CRC_CCITT16 = 0x1021;
	
	private static Hashtable<Integer, Class<?>>  m_vt2cls;
	private static Hashtable<Integer, int[]>     m_vt2opt;
	private static Hashtable<Class<?>, Class<?>> m_autobox;
	
	static {
		m_vt2cls = new Hashtable<Integer, Class<?>>();
		
		m_vt2cls.put(VT_UI1    , Byte.class      );
		m_vt2cls.put(VT_I2     , Short.class     );
		m_vt2cls.put(VT_UI2    , Short.class     );
		m_vt2cls.put(VT_I4     , Integer.class   );
		m_vt2cls.put(VT_UI4    , Integer.class   );
		m_vt2cls.put(VT_I8     , Long.class      );
		m_vt2cls.put(VT_UI8    , Long.class      );		
		m_vt2cls.put(VT_ERROR  , Integer.class   );
		m_vt2cls.put(VT_R4     , Float.class     );
		m_vt2cls.put(VT_R8     , Double.class    );
		m_vt2cls.put(VT_CY     , BigDecimal.class);
		m_vt2cls.put(VT_DATE   , Date.class      );
		m_vt2cls.put(VT_BOOL   , Boolean.class   );
		m_vt2cls.put(VT_BSTR   , String.class    );
		
		// Array
		m_vt2cls.put(VT_ARRAY | VT_UI1    , byte[].class      );
		m_vt2cls.put(VT_ARRAY | VT_I2     , short[].class     );
		m_vt2cls.put(VT_ARRAY | VT_UI2    , short[].class     );
		m_vt2cls.put(VT_ARRAY | VT_I4     , int[].class       );
		m_vt2cls.put(VT_ARRAY | VT_UI4    , int[].class       );
		m_vt2cls.put(VT_ARRAY | VT_I8     , long[].class      );
		m_vt2cls.put(VT_ARRAY | VT_UI8    , long[].class      );
		m_vt2cls.put(VT_ARRAY | VT_R4     , float[].class     );
		m_vt2cls.put(VT_ARRAY | VT_R8     , double[].class    );
		m_vt2cls.put(VT_ARRAY | VT_CY     , BigDecimal[].class);
		m_vt2cls.put(VT_ARRAY | VT_DATE   , Date[].class      );
		m_vt2cls.put(VT_ARRAY | VT_BOOL   , boolean[].class   );
		m_vt2cls.put(VT_ARRAY | VT_BSTR   , String[].class    );
		m_vt2cls.put(VT_ARRAY | VT_VARIANT, Object[].class    );
		
		// int[2] -> [0]: the number of bytes, [1]: primary
		m_vt2opt = new Hashtable<Integer, int[]>();
		m_vt2opt.put(VT_UI1    , new int[]{1,1});
		m_vt2opt.put(VT_I2     , new int[]{2,1});
		m_vt2opt.put(VT_UI2    , new int[]{2,0});
		m_vt2opt.put(VT_I4     , new int[]{4,1});
		m_vt2opt.put(VT_UI4    , new int[]{4,0});
		m_vt2opt.put(VT_I8     , new int[]{8,1});
		m_vt2opt.put(VT_UI8    , new int[]{8,0});
		m_vt2opt.put(VT_ERROR  , new int[]{4,0});
		m_vt2opt.put(VT_R4     , new int[]{4,1});
		m_vt2opt.put(VT_R8     , new int[]{8,1});
		m_vt2opt.put(VT_CY     , new int[]{8,1});
		m_vt2opt.put(VT_DATE   , new int[]{8,1});
		m_vt2opt.put(VT_BOOL   , new int[]{2,1});
		m_vt2opt.put(VT_BSTR   , new int[]{4,1});
		m_vt2opt.put(VT_VARIANT, new int[]{0,1});
		
		m_autobox = new Hashtable<Class<?>, Class<?>>();
		m_autobox.put(Byte[].class   , byte[].class   );		
		m_autobox.put(Short[].class  , short[].class  );
		m_autobox.put(Integer[].class, int[].class    );
		m_autobox.put(Float[].class  , float[].class  );
		m_autobox.put(Double[].class , double[].class );
		m_autobox.put(Long[].class   , long[].class   );
		m_autobox.put(Boolean[].class, boolean[].class);
	}
	
	private static int FindVtFromClass(Class<?> cls) throws Throwable {
		int ret = -1, cnt = 0;
		// cnt=0: original class, =1: boxing class, >1: break
		while(ret == -1) {
			// if did not find from original class, then find from boxing class
			if(cnt == 1) {
				cls = m_autobox.get(cls);
			} else if(cnt > 1) {
				break;
			}
			
			for(Entry<Integer, Class<?>> e : m_vt2cls.entrySet()) {
				if(cls.equals(e.getValue())) {
					int tmp = e.getKey();
					int[] opt = m_vt2opt.get(tmp & ~VT_ARRAY);
					// if the found vt is primary, then return that
					if(opt[1] == 1) {
						ret = tmp;
						break;
					}
				}
			}
			
			cnt++;
		}
		
		return ret;
	}
	
	private boolean m_hasCRC;
	
	public BCAPPacketConverter() {
		this(false);
	}
	
	public BCAPPacketConverter(boolean hasCRC) {
		m_hasCRC = hasCRC;
	}
	
	public byte[] Serialize(BCAPPacket packet) throws IllegalArgumentException {
		ByteArrayOutputStream bs = new ByteArrayOutputStream();
		byte[] b = null;
		
		try {
			bs.write(BCAP_SOH);
			bs.write(new byte[] {0, 0, 0, 0});
			bs.write(BCAPByteConverter.Object2Byte(packet.Serial));
			bs.write(BCAPByteConverter.Object2Byte(packet.Reserv));
			bs.write(BCAPByteConverter.Object2Byte(packet.FuncID));
			bs.write(BCAPByteConverter.Object2Byte((short)packet.Args.size()));
			Serialize_ObjectArray(bs, packet.Args, true);
			if(m_hasCRC) {
				bs.write(new byte[] {0, 0});
			}
			bs.write(BCAP_EOT);
			
			b = bs.toByteArray();
			ByteBuffer.wrap(b, 1, 4).put(BCAPByteConverter.Object2Byte(b.length));
			if(m_hasCRC) {
				ByteBuffer.wrap(b, b.length - 3, 2).put(
						BCAPByteConverter.Object2Byte((short)CalcCRC16(b, 1, b.length - 4)));
			}
		} catch(Throwable cause) {
			throw new IllegalArgumentException(cause);
		}
		
		return b;
	}
	
	private void Serialize_ObjectArray(ByteArrayOutputStream bs, ArrayList<Object> args, boolean first) throws Throwable {
		for(Object obj : args) {
			ByteArrayOutputStream _bs = new ByteArrayOutputStream();
						
			if (first) {
				_bs.write(new byte[] {0, 0, 0, 0});
			}
			
			Serialize_Object(_bs, obj);

			byte[] b = _bs.toByteArray();
			if (first) {
				ByteBuffer.wrap(b, 0, 4).put(BCAPByteConverter.Object2Byte(b.length - 4));
			}
			
			bs.write(b);
		}
	}
	
	private void Serialize_Object(ByteArrayOutputStream bs, Object arg) throws Throwable {
		if(arg == null) {
			bs.write(BCAPByteConverter.Object2Byte((short)VT_EMPTY));
			bs.write(BCAPByteConverter.Object2Byte(1));
			return;
		}
		
		Class<?> cls = arg.getClass();
		int vt = FindVtFromClass(cls);
		bs.write(BCAPByteConverter.Object2Byte((short)vt));
		
		if(cls.isArray()) {
			int argc = Array.getLength(arg);
			bs.write(BCAPByteConverter.Object2Byte(argc));
			
			switch(vt ^ VT_ARRAY) {
				case VT_UI1:
					if(cls.equals(byte[].class)) {
						bs.write((byte[])arg);
					} else {
						for(int i = 0; i < argc; i++) {
							bs.write((Byte)Array.get(arg, i));
						}
					}
					break;
				case VT_BSTR:
					for(int i = 0; i < argc; i++) {
						String str = (String)Array.get(arg, i);
						bs.write(BCAPByteConverter.Object2Byte(2 * str.length()));
						if(!str.equals("")) {
							bs.write(BCAPByteConverter.Object2Byte(str));
						}
					}
					break;
				case VT_VARIANT:
					ArrayList<Object> ary = new ArrayList<Object>(argc);
					for(int i = 0; i < argc; i++) {
						ary.add(Array.get(arg, i));
					}
					Serialize_ObjectArray(bs, ary, false);
					break;
				default:
					for(int i = 0; i < argc; i++) {
						bs.write(BCAPByteConverter.Object2Byte(Array.get(arg, i)));
					}
					break;
			}
		} else {
			bs.write(BCAPByteConverter.Object2Byte(1));
			
			switch(vt) {
				case VT_UI1:
					bs.write((Byte)arg);
					break;
				case VT_BSTR:
					String str = (String)arg;
					bs.write(BCAPByteConverter.Object2Byte(2 * str.length()));
					if(!str.equals("")) {
						bs.write(BCAPByteConverter.Object2Byte(str));
					}
					break;
				default:
					bs.write(BCAPByteConverter.Object2Byte(arg));
					break;
			}
		}
	}
	
	public BCAPPacket Deserialize(byte[] packet) throws IllegalArgumentException {
		BCAPPacket ret = new BCAPPacket();

		try {
			ByteArrayInputStream bs = new ByteArrayInputStream(packet);
			byte[] b = new byte[MAX_OPT_BYTES];
		
			int soh = bs.read(); // SOH
			
			bs.read(b, 0, 4); // Length
			int len = (Integer)BCAPByteConverter.Byte2Object(b, Integer.class);
			if(soh != BCAP_SOH || len > packet.length || packet[len - 1] != BCAP_EOT) {
				throw new IllegalArgumentException("The header or terminator is wrong.");
			}
			
			bs.read(b, 0, 2); // Serial
			ret.Serial = (Short)BCAPByteConverter.Byte2Object(b, Short.class);
			
			bs.read(b, 0, 2); // Reservation
			ret.Reserv = (Short)BCAPByteConverter.Byte2Object(b, Short.class);
			
			bs.read(b, 0, 4); // Function ID
			ret.FuncID = (Integer)BCAPByteConverter.Byte2Object(b, Integer.class);
			
			bs.read(b, 0, 2); // Number of arguments
			int argc = (Short)BCAPByteConverter.Byte2Object(b, Short.class);
			
			ret.Args = Deserialize_ObjectArray(bs, argc, true);
			
			if(m_hasCRC) {
				short crc_calc, crc_recv;
				crc_calc = (short)CalcCRC16(packet, 1, len - 4);
				bs.read(b, 0, 2);
				crc_recv = (Short)BCAPByteConverter.Byte2Object(b, Short.class);
				if(crc_calc != crc_recv) {
					throw new IllegalArgumentException("CRC is wrong.");
				}
			}
			
			bs.read(); // EOH
		} catch(Throwable cause) {
			throw new IllegalArgumentException(cause);
		}
		
		return ret;
	}
	
	private ArrayList<Object> Deserialize_ObjectArray(ByteArrayInputStream bs, int argc, boolean first) throws Throwable {
		ArrayList<Object> args = new ArrayList<Object>(argc);
		
		for(int i = 0; i < argc; i++) {
			if(first) {
				bs.skip(4);
			}
			
			args.add(Deserialize_Object(bs));
		}
		
		return args;
	}
	
	private Object Deserialize_Object(ByteArrayInputStream bs) throws Throwable {
		Object obj = null;
		byte[] bTmp, b = new byte[MAX_OPT_BYTES];
		
		bs.read(b, 0, 2);
		int vt = (Short)BCAPByteConverter.Byte2Object(b, Short.class);
		
		bs.read(b, 0, 4);
		int argc = (Integer)BCAPByteConverter.Byte2Object(b, Integer.class);
		
		if((vt & VT_ARRAY) != 0) {
			Class<?> cls  = m_vt2cls.get(vt).getComponentType();
			int[] opt = m_vt2opt.get(vt ^ VT_ARRAY);

			switch(vt ^ VT_ARRAY) {
				case VT_UI1:
					bTmp = new byte[argc];
					bs.read(bTmp, 0, argc);
					obj = bTmp;
					break;
				case VT_BSTR:
					obj = Array.newInstance(String.class, argc);
					for(int i = 0; i < argc; i++) {
						bs.read(b, 0, opt[0]);
						int len = (Integer)BCAPByteConverter.Byte2Object(b, Integer.class);
						if(len > 0) {
							bTmp = new byte[len];
							bs.read(bTmp, 0, len);
							Array.set(obj, i, BCAPByteConverter.Byte2Object(bTmp, String.class));
						} else Array.set(obj, i, "");
					}
					break;
				case VT_VARIANT:
					ArrayList<Object> ary = Deserialize_ObjectArray(bs, argc, false);
					obj = ary.toArray();
					break;
				default:
					obj = Array.newInstance(cls, argc);
					for(int i = 0; i < argc; i++) {
						bs.read(b, 0, opt[0]);
						Array.set(obj, i, BCAPByteConverter.Byte2Object(b, cls));
					}
					break;
			}
		} else {
			Class<?> cls = m_vt2cls.get(vt);
			int[] opt = m_vt2opt.get(vt);
			
			switch(vt) {
				case VT_EMPTY:
				case VT_NULL:
					break;
				case VT_UI1:
					obj = new Byte((byte)bs.read());
					break;
				case VT_BSTR:
					bs.read(b, 0, opt[0]);
					int len = (Integer)BCAPByteConverter.Byte2Object(b, Integer.class);
					if(len > 0) {
						bTmp = new byte[len];
						bs.read(bTmp, 0, len);
						obj = BCAPByteConverter.Byte2Object(bTmp, String.class);
					} else obj = "";
					break;
				default:
					bs.read(b, 0, opt[0]);
					obj = BCAPByteConverter.Byte2Object(b, cls);
					break;
			}
		}
		
		return obj;
	}
	
	private int CalcCRC16(byte[] bin, int offset, int length){
		int i, j, crc = 0xFFFF;
		
		for(i = offset; i < offset + length; i++){
			crc ^= (bin[i] << 8);
			for(j = 0; j < 8; j++){
				if((crc & 0x8000) != 0){
					crc <<= 1;
					crc ^= CRC_CCITT16;
				}else{
					crc <<= 1;
				}
			}
		}
		
		return crc;
	}	
}

File: CSharp/bCAPClient/Src/bCAPPacketConverter.cs
/**
 * Software License Agreement (MIT License)
 *
 * @copyright Copyright (c) 2015 DENSO WAVE INCORPORATED
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.IO;
using System.Text;
using ORiN2.Library;

namespace ORiN2.bCAP
{
    enum VARENUM : short
    {
        VT_EMPTY = 0,
        VT_NULL = 1,
        VT_I2 = 2,
        VT_I4 = 3,
        VT_R4 = 4,
        VT_R8 = 5,
        VT_CY = 6,
        VT_DATE = 7,
        VT_BSTR = 8,
        VT_ERROR = 10,
        VT_BOOL = 11,
        VT_VARIANT = 12,
        VT_UI1 = 17,
        VT_UI2 = 18,
        VT_UI4 = 19,
        VT_I8 = 20,
        VT_UI8 = 21,
        VT_ARRAY = 0x2000,
    }
    
    class bCAPPacketConverter
    {
        private const byte BCAP_SOH = 1;
        private const byte BCAP_EOT = 4;

        private static Dictionary<VARENUM, KeyValuePair<Type, int>> m_vt2type;

        static bCAPPacketConverter()
        {
            m_vt2type = new Dictionary<VARENUM, KeyValuePair<Type, int>>();
            m_vt2type[VARENUM.VT_I2] = new KeyValuePair<Type, int>(typeof(Int16), 2);
            m_vt2type[VARENUM.VT_I4] = new KeyValuePair<Type, int>(typeof(Int32), 4);
            m_vt2type[VARENUM.VT_R4] = new KeyValuePair<Type, int>(typeof(Single), 4);
            m_vt2type[VARENUM.VT_R8] = new KeyValuePair<Type, int>(typeof(Double), 8);
            m_vt2type[VARENUM.VT_CY] = new KeyValuePair<Type, int>(typeof(Decimal), 8);
            m_vt2type[VARENUM.VT_DATE] = new KeyValuePair<Type, int>(typeof(DateTime), 8);
            m_vt2type[VARENUM.VT_BSTR] = new KeyValuePair<Type, int>(typeof(String), -1);
            m_vt2type[VARENUM.VT_BOOL] = new KeyValuePair<Type, int>(typeof(Boolean), 2);
            m_vt2type[VARENUM.VT_UI1] = new KeyValuePair<Type, int>(typeof(Byte), 1);
            m_vt2type[VARENUM.VT_UI2] = new KeyValuePair<Type, int>(typeof(UInt16), 2);
            m_vt2type[VARENUM.VT_UI4] = new KeyValuePair<Type, int>(typeof(UInt32), 4);
            m_vt2type[VARENUM.VT_I8] = new KeyValuePair<Type, int>(typeof(Int64), 8);
            m_vt2type[VARENUM.VT_UI8] = new KeyValuePair<Type, int>(typeof(UInt64), 8);
            // 検索避けのためTypeが被る型は後で追加
            m_vt2type[VARENUM.VT_EMPTY] = new KeyValuePair<Type, int>(null, 0);
            m_vt2type[VARENUM.VT_NULL] = new KeyValuePair<Type, int>(null, 0);
            m_vt2type[VARENUM.VT_ERROR] = new KeyValuePair<Type, int>(typeof(Int32), 4);
        }

        /* Typeをキーに型を検索 */
        private static VARENUM FindVtFromType(Type t)
        {
            return m_vt2type.First(e => e.Value.Key == t).Key;
        }

        public byte[] Serialize(bCAPPacket packet)
        {
            byte[] bRet = null;

            try
            {
                using (var ms = new MemoryStream())
                {
                    using (var bw = new BinaryWriter(ms))
                    {
                        // SOH
                        bw.Write(BCAP_SOH);

                        // Message Length（最後に正しい値を入力）
                        bw.Write((Int32)0);

                        // Serial Number
                        bw.Write(bCAPByteConverter.Value2Byte(packet.iSerial));

                        // Reserve Number
                        bw.Write(bCAPByteConverter.Value2Byte(packet.iReserv));

                        // Function ID
                        bw.Write(bCAPByteConverter.Value2Byte(packet.iFuncID));

                        // Number of arguments
                        bw.Write(bCAPByteConverter.Value2Byte((Int16)packet.aryArgs.Count));

                        // Arguments
                        Serialize_ObjectArray(bw, packet.aryArgs, true);

                        // EOH
                        bw.Write(BCAP_EOT);

                        // Message Length（正しい値を入力）
                        bw.BaseStream.Position = 1;
                        bw.Write(bCAPByteConverter.Value2Byte((Int32)ms.Length));
                    }

                    bRet = ms.ToArray();
                }
            }
            catch (Exception ex)
            {
                throw new ORiN2Exception(HResult.E_INVALIDARG, ex);
            }

            return bRet;
        }

        private void Serialize_ObjectArray(BinaryWriter bw, ArrayList args, bool first = false)
        {
            long lStart, lEnd;

            foreach (object obj in args)
            {
                // 変換前のオフセットを記憶
                lStart = bw.BaseStream.Position;

                if (first)
                {
                    // 引数サイズ（最後に正しい値を入力）
                    bw.Write((Int32)0);
                }

                // 引数
                Serialize_Object(bw, obj);

                // 変換後のオフセットを記憶
                lEnd = bw.BaseStream.Position;

                if (first)
                {
                    // 引数サイズ（正しい値を入力）
                    bw.BaseStream.Position = lStart;
                    bw.Write(bCAPByteConverter.Value2Byte((Int32)(lEnd - lStart - 4)));

                    // 最後尾に戻す
                    bw.BaseStream.Position = lEnd;
                }
            }
        }

        private void Serialize_Object(BinaryWriter bw, object obj)
        {
            if (obj == null)
            {
                // 型: VT_EMPTY, 要素数: 1
                bw.Write(bCAPByteConverter.Value2Byte((Int16)VARENUM.VT_EMPTY));
                bw.Write(bCAPByteConverter.Value2Byte((Int32)1));
            }
            else
            {
                // 引数の型を取得
                Type t = obj.GetType();

                if (t.IsArray)
                {
                    // 配列に変換
                    Array objArray = (Array)obj;

                    // 要素の型を取得
                    t = t.GetElementType();

                    if(t == typeof(Object))
                    {
                        // 型: (VT_VARIANT | VT_ARRAY), 要素数: objArray.Length
                        bw.Write(bCAPByteConverter.Value2Byte((Int16)(VARENUM.VT_VARIANT | VARENUM.VT_ARRAY)));
                        bw.Write(bCAPByteConverter.Value2Byte((Int32)objArray.Length));
                        // 再帰
                        Serialize_ObjectArray(bw, new ArrayList(objArray));
                    }
                    else
                    {
                        // TypeをVARENUMに変換
                        VARENUM vt = FindVtFromType(t);

                        // 型: (vt | VT_ARRAY), 要素数: objArray.Length
                        bw.Write(bCAPByteConverter.Value2Byte((Int16)(vt | VARENUM.VT_ARRAY)));
                        bw.Write(bCAPByteConverter.Value2Byte((Int32)objArray.Length));

                        if (t == typeof(Byte))
                        {
                            // バイト配列はそのまま代入
                            bw.Write((byte[])objArray);
                        }
                        else if (t == typeof(String))
                        {
                            // 文字列配列をバイト配列に変換
                            foreach (String tmp in objArray)
                            {
                                bw.Write(bCAPByteConverter.Value2Byte((Int32)(2 * tmp.Length)));
                                bw.Write(bCAPByteConverter.Value2Byte(tmp));
                            }
                        }
                        else
                        {
                            // 各要素をバイト配列に変換
                            foreach (object tmp in objArray)
                            {
                                bw.Write(bCAPByteConverter.Value2Byte(tmp));
                            }
                        }
                    }
                }
                else
                {
                    // TypeをVARENUMに変換
                    VARENUM vt = FindVtFromType(t);

                    // 型: vt, 要素数: 1
                    bw.Write(bCAPByteConverter.Value2Byte((Int16)vt));
                    bw.Write(bCAPByteConverter.Value2Byte((Int32)1));

                    if (t == typeof(Byte))
                    {
                        // バイトはそのまま代入
                        bw.Write((byte)obj);
                    }
                    else if (t == typeof(String))
                    {
                        // 文字列をバイト配列に変換
                        bw.Write(bCAPByteConverter.Value2Byte((Int32)(2 * ((String)obj).Length)));
                        bw.Write(bCAPByteConverter.Value2Byte(obj));
                    }
                    else
                    {
                        // 要素をバイト配列に変換
                        bw.Write(bCAPByteConverter.Value2Byte(obj));
                    }
                }
            }
        }

        public bCAPPacket Deserialize(byte[] buf)
        {
            var packet = new bCAPPacket();

            try
            {
                using (var ms = new MemoryStream(buf))
                {
                    using (var br = new BinaryReader(ms))
                    {
                        // Message Length
                        br.BaseStream.Position = 1;
                        Int32 len = (Int32)bCAPByteConverter.Byte2Value(br.ReadBytes(4), typeof(Int32));

                        // パケットの確認
                        if ((buf[0] != BCAP_SOH) ||
                            (len > buf.Length) || (buf[len - 1] != BCAP_EOT))
                        {
                            throw new ArgumentException();
                        }

                        // Serial Number
                        packet.iSerial = (UInt16)bCAPByteConverter.Byte2Value(br.ReadBytes(2), typeof(UInt16));

                        // Reserve Number
                        packet.iReserv = (UInt16)bCAPByteConverter.Byte2Value(br.ReadBytes(2), typeof(UInt16));

                        // Function ID
                        packet.iFuncID = (Int32)bCAPByteConverter.Byte2Value(br.ReadBytes(4), typeof(Int32));

                        // Number of arguments
                        Int16 argc = (Int16)bCAPByteConverter.Byte2Value(br.ReadBytes(2), typeof(Int16));

                        // Arguments
                        packet.aryArgs.AddRange(Deserialize_ObjectArray(br, argc, true));
                    }
                }
            }
            catch (Exception ex)
            {
                throw new ORiN2Exception(HResult.E_INVALIDPACKET, ex);
            }

            return packet;
        }

        private ArrayList Deserialize_ObjectArray(BinaryReader br, int argc, bool first = false)
        {
            ArrayList args = new ArrayList(argc);

            for (int i = 0; i < argc; i++)
            {
                if (first)
                {
                    // 引数サイズをスキップ
                    br.BaseStream.Position += 4;
                }

                // 引数を変換して追加
                args.Add(Deserialize_Object(br));
            }

            return args;
        }

        private object Deserialize_Object(BinaryReader br)
        {
            object obj = null;

            // 引数の型
            Int16 vt = (Int16)bCAPByteConverter.Byte2Value(br.ReadBytes(2), typeof(Int16));
            
            // 要素数
            Int32 argc = (Int32)bCAPByteConverter.Byte2Value(br.ReadBytes(4), typeof(Int32));

            if ((vt & (Int16)VARENUM.VT_ARRAY) != 0)
            {
                vt ^= (Int16)VARENUM.VT_ARRAY;

                if (vt == (Int16)VARENUM.VT_VARIANT)
                {
                    // 型: (VT_VARIANT | VT_ARRAY), 要素数: argc
                    ArrayList objArray = Deserialize_ObjectArray(br, argc);
                    obj = objArray.ToArray();
                }
                else
                {
                    var pair = m_vt2type[(VARENUM)vt];

                    // VARENUMをTypeに変換
                    Type t = pair.Key;

                    // Typeのバイト数を取得
                    Int32 len = pair.Value;

                    if (t == typeof(Byte))
                    {
                        // バイト配列はそのまま代入
                        obj = br.ReadBytes(argc);
                    }
                    else if (t == typeof(String))
                    {
                        // バイト配列を文字列配列に変換
                        String[] strArray = new String[argc];
                        for (int i = 0; i < argc; i++)
                        {
                            len = (Int32)bCAPByteConverter.Byte2Value(br.ReadBytes(4), typeof(Int32));
                            strArray[i] = (String)bCAPByteConverter.Byte2Value(br.ReadBytes(len), t);
                        }
                        obj = strArray;
                    }
                    else
                    {
                        // バイト配列を各要素に変換
                        ArrayList objArray = new ArrayList(argc);
                        for (int i = 0; i < argc; i++)
                        {
                            objArray.Add(bCAPByteConverter.Byte2Value(br.ReadBytes(len), t));
                        }
                        obj = objArray.ToArray(t);
                    }
                }
            }
            else
            {
                var pair = m_vt2type[(VARENUM)vt];

                // VARENUMをTypeに変換
                Type t = pair.Key;

                // Typeのバイト数を取得
                Int32 len = pair.Value;

                if (t == null)
                {
                    obj = null;
                }
                else if (t == typeof(Byte))
                {
                    // バイトはそのまま代入
                    obj = br.ReadByte();
                }
                else if (t == typeof(String))
                {
                    // バイト配列を文字列に変換
                    len = (Int32)bCAPByteConverter.Byte2Value(br.ReadBytes(4), typeof(Int32));
                    obj = bCAPByteConverter.Byte2Value(br.ReadBytes(len), t);
                }
                else
                {
                    // バイト配列を要素に変換
                    obj = bCAPByteConverter.Byte2Value(br.ReadBytes(len), t);
                }
            }

            return obj;
        }
    }
}
